### PR TITLE
map returns copy

### PR DESCRIPTION
**Pull request by adamjanicki2**
addresses #15

### CODE REVIEW

**Comment by LeaVerou** [comment]
I'm not quite sure what `childrenWithProperty()` does, and why we need it. Could you elaborate?

**Comment by LeaVerou** [comment]
I see. I don't think it's a good idea to add that export to children.js, since I don't think it's more broadly useful. Even if we did need it, it's probably more suited to a private function in the map module. That said, I'm not quite sure we need it, each invocation of `_map()` includes a pointer to the parent AND the property, and that's where we set `parent[property]` as well, so I'd just clone `node` if there was no result, and set `parent[property]` anyway.

**Comment by LeaVerou** [comment]
Quasi-related: I wonder if `transform()` might be a better name than `map()`. I chose `map()` as it relates to an existing concept authors may be familiar with, but if it doesn't have enough similarities that could hinder, rather than improve usability.

**Comment by adamjanicki2** [comment]
> Quasi-related: I wonder if `transform()` might be a better name than `map()`. I chose `map()` as it relates to an existing concept authors may be familiar with, but if it doesn't have enough similarities that could hinder, rather than improve usability.

I definitely support this, to me, when I think of `map`, I think of something linear, not a complex tree structure like we have

**Review by LeaVerou — CHANGES_REQUESTED** [review]
(Flushing the comments I had already written)

**Comment by LeaVerou** [comment]
Let's merge this, the other issue is entirely orthogonal!

File: src/children.js [hunk 1]
```diff
@@ -28,4 +28,4 @@ export const properties = children.properties = {
 };
 
 // Old JSEP versions
-childProperties.LogicalExpression = childProperties.BinaryExpression;
\ No newline at end of file
+properties.LogicalExpression = properties.BinaryExpression;
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/children.js:31]: You didn't port the changes manually, right? You're supposed to rebase etc

File: src/map.js [hunk 2]
```diff
@@ -15,21 +16,23 @@ export default function map (node, callback, o) {
...
-	let ignored = o.except && matches(node, o.except);
+	if (Array.isArray(node)) {
+		return node.map(n => _map(n, callback, o, property, parent));
+	}
 
-	if (!ignored && matches(node, o.only)) {
-		let ret = callback(node, property, parent);
+	const ignore = o.except && matches(node, o.except);
+	const explore = !ignore && matches(node, o.only);
+	let copiedNode = {...node};
```
Comment (LeaVerou) [inline src/map.js:25]: Why did you move this here? I had it where I had it so that we don't clone unless it's actually needed.
Btw if you're ok with my changes, instead of porting them here and I try to figure out what you changed and why, another option would be for you to push the tests to my branch (or merge and reduce this PR to the tests and the other fix).
